File: content/tracks/dsa/problems/lc-0424-longest-repeating-character-replacement/solution.py
from typing import List


class Solution:
    def characterReplacement(self, s: str, k: int) -> int:
        # Sliding window: the window is valid while (length - count of its most frequent letter)
        # <= k. The best count seen so far never needs to shrink: only a larger count can give a
        # longer answer.
        counts: List[int] = [0] * 26
        left = 0
        top = 0
        best = 0
        for right, char in enumerate(s):
            index = ord(char) - ord("A")
            counts[index] += 1
            top = max(top, counts[index])
            while right - left + 1 - top > k:
                counts[ord(s[left]) - ord("A")] -= 1
                left += 1
            best = max(best, right - left + 1)
        return best

File: content/tracks/dsa/problems/lc-0424-longest-repeating-character-replacement/solution.go
package main

func characterReplacement(s string, k int) int {
	// Sliding window: the window is valid while (length - count of its most frequent letter)
	// <= k. The best count seen so far never needs to shrink: only a larger count can give a
	// longer answer.
	var counts [26]int
	left, top, best := 0, 0, 0
	for right := 0; right < len(s); right++ {
		index := s[right] - 'A'
		counts[index]++
		top = max(top, counts[index])
		for right-left+1-top > k {
			counts[s[left]-'A']--
			left++
		}
		best = max(best, right-left+1)
	}
	return best
}
